### ISSUE TITLE
Paths matching `*` are replaced again when `resolve-tspaths` is run more than once

### ISSUE THREAD

**Issue by wintercounter** [benyap/resolve-tspaths#122]
`require("../..")` becomes `require("../../../../..")`

Try to rerun `resolve-tspaths` multiple times on the same files.

**Comment by benyap** [comment]
Hi @wintercounter, could you please provide clear steps to reproduce the issue?

**Comment by wintercounter** [comment]
First I couldn't repro, then I figured out. In `tsconfig` I have a `*` key under `paths`, that's causing the problem. I needed that because in my projects I have multiple `node_modules` folders, don't ask why :)

I created a repro here, please open it in Gitpod: https://github.com/wintercounter/replace-tspaths-repro

Steps:
1. Run `npm i`
2. Run `npm run build`
3. Check `dist/processes/ecosystem/root/index.js`, it should have `require('@/processes')`
4. Run `npm run replace`
5. Check `dist/processes/ecosystem/root/index.js` again, you should see `require('../..')`
6. Run `npm run replace` again,  _WITHOUT_ new build.
7. The same require statement will become `require("../../../../..")`.

**Comment by benyap** [comment]
Hi @wintercounter, I took a look at your issue. Thanks for providing the reproduction repo.

- You're definitely right in that it's the `*` key in your tsconfig that's causing the behaviour that you're seeing.
- Unfortunately though, `resolve-tspaths` cannot promise to be [idempotent](https://stackoverflow.com/questions/1077412/what-is-an-idempotent-operation) when you use `*`.
- When `resolve-tspaths` scans your output directory, there's no way it can tell whether a `*` path has been replaced since it matches everything, unlike a path with a prefix.
- I might have some solutions for you though, which I'll detail further below.

I'm going to close this issue as there isn't any fix required. Would love to hear your feedback though and if you're facing issues, please feel free to re-open. I'm also going to rename the issue to make it a little clearer so others can find this if they come across this same problem.

### Potential solutions

#### 1. Run `resolve-tspath` as a part of your build step

This package was designed to be run as a part of your build step (i.e. to be run after tsc/swc/esbuild etc. does its thing), as opposed to being a command you run manually, or something you run multiple times (though it has no effect unless you have a `*` path as you have discovered). So my recommendation would be to make your `build` script look something like below so that it's run automatically after a fresh build only.
```
swc src --out-dir dist --ignore dist,node_modules && resolve-tspaths -o dist -s src 
```
I haven't come across a use case where you would need to run `resolve-tspaths` more than once after a build. If this is happens to be your use case though, please let me know.

#### 2. Create an additional `tsconfig` file for use with `resolve-tspaths` without `*`

I'm making the assumption here that you don't actually need paths that match `*` to be replaced by `resolve-tspaths` since you mention they're for imports pointing to node_modules folders. If this is the case, you could make a new `tsconfig.build.json` file that extends `tsconfig.json`, but override the `paths` block to only have your other aliases and not the one with `*`:
```jsonc
// tsconfig.build.json
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    // ...
    "paths": {
        // path for * is removed
        "@": ["/workspace/replace-tspaths-repro/src/index"],
        "@/*": ["/workspace/replace-tspaths-repro/src/*"]
    }
  }  
}
```
And pass this file to resolve-tspaths explicitly so that it won't try to replace paths that match `*`:
```
resolve-tspaths -o dist -s src -p tsconfig.build.json
```
The cost of this solution is that you need to make sure that the other paths in `tsconfig.json` and `tsconfig.build.json` are in sync.

**Comment by wintercounter** [comment]
I didn't have this problem with `tscpaths`. The issue is that I'm running this as part of a build tool with an swc watcher. So basically after each swc change `resolve-tspaths` will be invoked. I cannot ask my users to not have `*` in their tsconfig in order to use my build tool...

As I saw, this is only an issue with non-named paths, can't it be just tested like `^[\\/\.]*$` and skip?

**Comment by benyap** [comment]
I see, that makes more sense! I'll look into your suggestion and get back to you.

**Comment by benyap** [comment]
This issue should be resolved in the [0.7.1](https://github.com/benyap/resolve-tspaths/releases/tag/v0.7.1) release.

Relative paths `./*` and `../*` will not be replaced by `resolve-tspaths`. A consequence of this is that you won't be able to use them as aliases, but that's a terrible idea anyway.

Also, I was curious as to why you didn't run into this problem with `tscpaths`, and turns out that it totally ignores the `*` alias - see [here](https://github.com/joonhocho/tscpaths/blob/master/src/index.ts#L81-L88), where the `*` alias would have been filtered out by line 88. I guess that's one way to do it haha, but definitely feels a little "wrong".
